### PR TITLE
APEXCORE-10 #resolve Changes for supporting anti-affinity in operators

### PR DESCRIPTION
**Pull request by ishark**
Added list of affinity rules as attribute in DAGContext
Added validation for affinity, anti-affinity and locality rules in dag validate phase
Added handling of Container and Node affinity rules. Rack affinity is not supported at this point
Also added Changes for supporting Node specific request in Cloudera
Added Unit tests for affinity rules and dag validation for the same
Added Json String codec for setting affinity rules through properties.xml
Added Set Affinity and anti -affinity APIs to dag

### CODE REVIEW

**Comment by ishark** [comment]
@tweise Pleas review

**Comment by ishark** [comment]
Testing Done: 
1. Unit tests for Dag Validation scenarios
2. Unit tests for host allocation based on affinity rules
3. Tested following scenarios on Cloudera and Hortonworks:
   - Anti-affinity between partitions of operators
   - Anti-affinity between any two operators (stream connected and not connected) 
   - Anti-affinity between parallel partitioned operators
   - Node and Container Affinity between operators not connected by stream
   - Node affinity between 2 Stream connected operators when the stream is not Node Local
   - Container affinity between 2 Stream connected operators when the stream is not Node Local
   - Affinity and Anti-affinity rules with Locality_host set for operator
   - Out of resources condition in case of Strict Anti-affinity. Application should keep waiting till resources are allocated as per rules
   - Checked preferred anti-affinity rules are dropped if resources are not available

**Comment by tweise** [comment]
@ishark there are merge conflicts, can you please rebase.

File: api/pom.xml [hunk 1]
```diff
@@ -70,14 +70,6 @@
           <groupId>commons-beanutils</groupId>
         </exclusion>
         <exclusion>
-          <artifactId>jackson-core-asl</artifactId>
```
Comment (tweise) [inline api/pom.xml:73]: Why this change?

File: api/src/main/java/com/datatorrent/api/AffinityRule.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  private static final long serialVersionUID = 107131504929875386L;
+
+  /**
+   * Pair of operator names to specify affinity rule
+   * The order of operators is not considered in this class
+   * i.e. OperatorPair("O1", "O2") is equal to OperatorPair("O2", "O1")
+   */
+  public static class OperatorPair implements Serializable
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/AffinityRule.java]: Why not use a generic pair (or extend it)?

Comment (tweise) [inline api/src/main/java/com/datatorrent/api/AffinityRule.java]: Why do we need pair when we already have collection?

File: api/src/main/java/com/datatorrent/api/DAG.java [hunk 3]
```diff
@@ -250,6 +250,24 @@
   public abstract <T> void setAttribute(Operator operator, Attribute<T> key, T value);
 
   /**
+   * Set affinity between operators
+   * @param locality
+   * @param relaxLocality
+   * @param first operator
+   * @param one or more operators
+   */
+  public abstract void setAffinity(Locality locality, boolean relaxLocality, String firstOperator, String... operators);
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DAG.java]: Why these additions? Why not specify the attribute directly?

File: api/pom.xml (hunk 1 above)
Comment (ishark) [inline api/pom.xml:73]: This was added for supporting setting of Affinity Rules from properties.xml. Added a JsonStringCodec for reading Affinity Rule Json..

File: api/src/main/java/com/datatorrent/api/DAG.java (hunk 3 above)
Comment (ishark) [inline api/src/main/java/com/datatorrent/api/DAG.java]: Yes, the affinity/anti-affinity can be set from attributes directly. These were added as convenience methods as per discussion on dev mailing list. Here are sample usages with these methods:
https://github.com/ishark/Apex-Samples/blob/master/affinity-example/src/main/java/com/example/affinity/ApplicationDagAPIs.java#L33-L34

Compared to when setting via attributes:
https://github.com/ishark/Apex-Samples/blob/master/affinity-example/src/main/java/com/example/affinity/ApplicationDagContext.java#L42-L54

Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DAG.java]: I see. IMO we should not bloat the core API with convenience or helper methods.

File: api/src/main/java/com/datatorrent/api/AffinityRule.java (hunk 2 above)
Comment (ishark) [inline api/src/main/java/com/datatorrent/api/AffinityRule.java]: Collection support is for convenience. Instead of saying anti affinity between operator pairs lets say A & B , B & C, with collection we can directly set anti-affinity between collection A, B, C. 
Internally, everything is converted into operator pairs, as this is easier to validate against stream locality and other affinity rules.

The OperatorPair class was created because Pair class inside apex-common is not available under api.

File: api/src/main/java/com/datatorrent/api/StringCodec.java [hunk 4]
```diff
@@ -377,4 +383,43 @@ public String toString(Class<? extends T> clazz)
...
+    @Override
+    public String toString(T pojo)
+    {
+      try {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.configure(DeserializationConfig.Feature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        ObjectWriter writer = mapper.writer();
+        return writer.writeValueAsString(pojo);
+      } catch (IOException e) {
+        DTThrowable.wrapIfChecked(e);
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StringCodec.java]: This is not specific to this PR but in general, should we consider using Throwables.propagate() instead?

File: api/src/main/java/com/datatorrent/api/AffinityRule.java (hunk 2 above)
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/AffinityRule.java]: It's fine to use pair internally, but maybe it is not needed in the API? varargs looks more appropriate.

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+public class ContainerRequestHandler
+{
+  protected static final int NUMBER_MISSED_HEARTBEATS = 30;
+
+  public void reissueContainerRequests(AMRMClient<ContainerRequest> amRmClient, Map<StreamingContainerAgent.ContainerStartRequest, MutablePair<Integer, ContainerRequest>> requestedResources, int loopCounter, ResourceRequestHandler resourceRequestor, List<ContainerRequest> containerRequests, List<ContainerRequest> removedContainerRequests)
+  {
+    if (!requestedResources.isEmpty()) {
+      // resourceRequestor.clearNodeMapping();
+      for (Map.Entry<StreamingContainerAgent.ContainerStartRequest, MutablePair<Integer, ContainerRequest>> entry : requestedResources.entrySet()) {
+        if ((loopCounter - entry.getValue().getKey()) > NUMBER_MISSED_HEARTBEATS) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java]: What happens otherwise?

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+import org.apache.hadoop.yarn.client.api.AMRMClient.ContainerRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.stram.StreamingContainerAgent.ContainerStartRequest;
+
+/**
+ * Handles creating container requests and issuing node-specific container
+ * requests by blacklisting specifically for cloudera
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java]: Maybe a bit more explanation and JIRA references?

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java (hunk 5 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java]: This flow actually did not change from previously. I just moved it in a separate class. If container request was not honored by Yarn within this time frame, we reissue the request. This class was added to provide an override for Cloudera to manage Node specific request. Will update the Jira with these details.

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java (hunk 6 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java]: Will add it in comments, mentioning the Jira here for reference: https://issues.apache.org/jira/browse/YARN-1412

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ */
+public class ContainerRequestHandler
+{
+  protected static final int NUMBER_MISSED_HEARTBEATS = 30;
+
+  public void reissueContainerRequests(AMRMClient<ContainerRequest> amRmClient, Map<StreamingContainerAgent.ContainerStartRequest, MutablePair<Integer, ContainerRequest>> requestedResources, int loopCounter, ResourceRequestHandler resourceRequestor, List<ContainerRequest> containerRequests, List<ContainerRequest> removedContainerRequests)
+  {
+    if (!requestedResources.isEmpty()) {
+      // resourceRequestor.clearNodeMapping();
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java]: remove comment

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java (hunk 5 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandler.java]: A line comment to that effect may be helpful (enhancement :-).

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    // Issue all host specific requests first
+    if (!hostSpecificRequestsMap.isEmpty() && requestedResources.isEmpty()) {
+      LOG.info("Issue Host specific requests first");
+      // Blacklist all the nodes and issue request for host specific
+      Entry<String, List<ContainerRequest>> set = hostSpecificRequestsMap.entrySet().iterator().next();
+      List<ContainerRequest> requests = set.getValue();
+      List<String> blacklistNodes = resourceRequestor.getNodesExceptHost(requests.get(0).getNodes());
+      amRmClient.updateBlacklist(blacklistNodes, requests.get(0).getNodes());
+      blacklistedNodesForHostSpecificRequests = blacklistNodes;
+      LOG.info("Sending {} request(s) after blacklisting nodes {} and removed host from request {}", requests.size(), blacklistNodes, requests.get(0).getNodes());
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java]: Maybe it is better to log blacklisting all but ... since the list of nodes can be very large.

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        containerRequests.add(newCr);
+        hostSpecificRequests.remove(cr);
+      }
+      hostSpecificRequestsMap.remove(set.getKey());
+    } else if (!requestedResources.isEmpty()) {
+      // Check if any requests timed out, create new requests in that case
+      recreateContainerRequest(requestedResources, loopCounter, resourceRequestor, removedContainerRequests);
+    } else {
+      if (blacklistedNodesForHostSpecificRequests != null) {
+        // Remove the nodes blacklisted nodes for host specific requests
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java]: failed to parse comment :-)

File: engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.stram.StreamingContainerAgent.ContainerStartRequest;
+
+/**
+ * Handles creating container requests and issuing node-specific container
+ * requests by blacklisting specifically for cloudera
+ */
+public class ContainerRequestHandlerCloudera extends ContainerRequestHandler
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ContainerRequestHandlerCloudera.java]: Perhaps instead name it so it indicates what it does (blacklist based) vs. why we have it.

File: engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java [hunk 11]
```diff
@@ -108,6 +113,23 @@ public void updateNodeReports(List<NodeReport> nodeReports)
     }
   }
 
+  public List<String> getNodesExceptHost(List<String> hostNames)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java:165]: Does hostNames always contain one element? The contains operation is expensive...

File: engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java [hunk 12]
```diff
@@ -153,11 +177,22 @@ public String getHost(ContainerStartRequest csr, boolean first)
 
     // the host requested didn't have the resources so looking for other hosts
     host = null;
+    List<String> antiHosts = new ArrayList<>();
+    List<String> antiPreferredHosts = new ArrayList<>();
+    if (!c.getStrictAntiPrefs().isEmpty()) {
+      // Check if containers are allocated already for the anti-affinity
+      // containers
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java]: extra line break - check IDE settings

File: engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java [hunk 13]
```diff
@@ -170,19 +205,69 @@ public String getHost(ContainerStartRequest csr, boolean first)
...
+        } else {
+          host = INVALID_HOST;
         }
       }
     }
+    LOG.info("Found host {}", host);
     return host;
   }
 
+  public void populateAntiHostList(PTContainer c, List<String> antiHosts)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java:280]: Add javadoc

File: engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java [hunk 14]
```diff
@@ -170,19 +205,69 @@ public String getHost(ContainerStartRequest csr, boolean first)
...
+        // Check if there is an anti-affinity with host locality
+        String antiHost = getAntiHostsList(container);
+        if (antiHost != null) {
+          antiHosts.add(antiHost);
+        }
+      }
+    }
+  }
+
+  public String getAntiHostsList(PTContainer container)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java]: javadoc

File: engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java [hunk 15]
```diff
@@ -170,19 +205,69 @@ public String getHost(ContainerStartRequest csr, boolean first)
...
+      }
+      if (grpObj.getHost() != null) {
+        host = grpObj.getHost();
+        return host;
+      }
+    }
+    return null;
+  }
+
+  public String assignHost(String host, List<String> antiHosts, List<String> antiPreferredHosts, HostOperatorSet grpObj, Set<PTOperator> nodeLocalSet, int aggrMemory, int vCores)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java:327]: javadoc

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java [hunk 16]
```diff
@@ -668,7 +671,10 @@ private void execute() throws YarnException, IOException
     int numReleasedContainers = 0;
     int nextRequestPriority = 0;
     ResourceRequestHandler resourceRequestor = new ResourceRequestHandler();
+    // Use override for container requestor in case of cloudera distribution,to handle host specific requests
+    ContainerRequestHandler containerRequestor = System.getenv().containsKey("CDH_HADOOP_BIN") ? new ContainerRequestHandlerCloudera() : new ContainerRequestHandler();
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Would it makes sense to combine ResourceRequestHandler and ContainerRequestHandler?

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java [hunk 17]
```diff
@@ -713,6 +717,11 @@ private void execute() throws YarnException, IOException
         expiryTime = StramUserLogin.refreshTokens(tokenLifeTime, FileUtils.getTempDirectoryPath(), applicationId, conf, hdfsKeyTabFile, credentials, rmAddress, true);
       }
 
+      if (System.currentTimeMillis() - nodeReportUpdateTime >= UPDATE_NODE_REPORTS_INTERVAL) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: We should see to not retrieve currentTimeMillis in one iteration (also see security stuff above).

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java [hunk 18]
```diff
@@ -2435,7 +2435,7 @@ private void setAttributes(List<? extends Conf> confs, Attribute.AttributeMap at
           else {
             if (processedAttributes.add(attribute)) {
               String val = e.getValue();
-              if (val.trim().charAt(0) == '{') {
+              if (val.trim().charAt(0) == '{' && !(attribute.codec instanceof JsonStringCodec)) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java:2438]: Why this if we already have a way to parse JSON?

File: engine/src/test/java/com/datatorrent/stram/AffinityRulesTest.java [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+
+    AffinityRulesSet ruleSet = new AffinityRulesSet();
+    // Valid case:
+    List<AffinityRule> rules = new ArrayList<>();
+    ruleSet.setAffinityRules(rules);
+    AffinityRule rule1 = new AffinityRule(Type.ANTI_AFFINITY, new OperatorPair("O2", "O2"), Locality.NODE_LOCAL, false);
+    rules.add(rule1);
+    dag.setAttribute(DAGContext.AFFINITY_RULES_SET, ruleSet);
+    dag.validate();
+    dag.getAttributes().put(com.datatorrent.api.Context.DAGContext.APPLICATION_PATH, new File("target", HostLocalTest.class.getName()).getAbsolutePath());
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/AffinityRulesTest.java]: see StramTestSupport.TestMeta

**Comment by tweise** [comment]
Haven't looked at the plan changes yet, rest looks good so far.

File: engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java (hunk 11 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/ResourceRequestHandler.java:165]: We assign a single host for each request. However, Yarn ContainerRequest can have list of nodes.. Added hostnames to set for contains check.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java (hunk 18 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlanConfiguration.java:2438]: Nested objects parsing was failing with default JSON parsing. It needed className for correct parsing, so added JsonStringCodec.

**Comment by ishark** [comment]
Have addressed most of the review comments except convenience APIs in DAG. Did quick sanity testing on Cloudera after changes.
@tweise Could you please review?

**Comment by tweise** [comment]
Looking at it.

File: api/src/main/java/com/datatorrent/api/AffinityRule.java [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  public AffinityRule(Type type, Locality locality, boolean relaxLocality)
+  {
+    this.type = type;
+    this.locality = locality;
+    this.setRelaxLocality(relaxLocality);
+  }
+
+  public AffinityRule(Type type, Locality locality, boolean relaxLocality, String... operators)
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/AffinityRule.java]: Does it require at least 2 operators? Is so, maybe make it String, String...

File: api/src/main/java/com/datatorrent/api/DAG.java (hunk 3 above)
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DAG.java]: Is still think these should be removed. Any other opinions?

File: api/src/main/java/com/datatorrent/api/AffinityRule.java (hunk 20 above)
Comment (ishark) [inline api/src/main/java/com/datatorrent/api/AffinityRule.java]: Changed it to String, String ..

File: engine/src/main/java/com/datatorrent/stram/BlacklistBasedResourceRequestHandler.java [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+    // Issue all host specific requests first
+    if (!hostSpecificRequestsMap.isEmpty() && requestedResources.isEmpty()) {
+      LOG.info("Issue Host specific requests first");
+      // Blacklist all the nodes and issue request for host specific
+      Entry<String, List<ContainerRequest>> set = hostSpecificRequestsMap.entrySet().iterator().next();
+      List<ContainerRequest> requests = set.getValue();
+      List<String> blacklistNodes = resourceRequestor.getNodesExceptHost(requests.get(0).getNodes());
+      amRmClient.updateBlacklist(blacklistNodes, requests.get(0).getNodes());
+      blacklistedNodesForHostSpecificRequests = blacklistNodes;
+      LOG.info("Sending {} request(s) after blacklisting nodes {} and removed host from request {}", requests.size(), blacklistNodes, requests.get(0).getNodes());
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/BlacklistBasedResourceRequestHandler.java]: I still have a concern with the logging of potentially very large collections of nodes (consider cluster size).

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java [hunk 22]
```diff
@@ -750,6 +752,11 @@ private void execute() throws YarnException, IOException
         expiryTime = StramUserLogin.refreshTokens(tokenLifeTime, FileUtils.getTempDirectoryPath(), applicationId, conf, hdfsKeyTabFile, credentials, rmAddress, true);
       }
 
+      if (System.currentTimeMillis() > nodeReportUpdateTime) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: Still see System.currentTimeMillis repeated within the loop. Please make a final variable out of it.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 23]
```diff
@@ -344,9 +346,67 @@ public PhysicalPlan(LogicalPlan dag, PlanContext ctx) {
         addLogicalOperator(n);
       }
     }
+    
+    inlinePrefs.prefs.clear();
+    localityPrefs.prefs.clear();
+    
+    // Add inlinePrefs and localityPreds for affinity rules
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: typo

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 24]
```diff
@@ -357,25 +417,51 @@ public PhysicalPlan(LogicalPlan dag, PlanContext ctx) {
           if (!container.operators.isEmpty()) {
             LOG.warn("Operator {} shares container without locality contraint due to insufficient resources.", oper);
           }
+          // TODO: Check if PT Operators conflict in anti-affinity, Pick the first container that does not conflict
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Why would this happen here and not during logical plan validation?

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 25]
```diff
@@ -386,6 +472,38 @@ public PhysicalPlan(LogicalPlan dag, PlanContext ctx) {
     this.undeployOpers.clear();
   }
 
+  public void setAntiAffinityForContainers(LogicalPlan dag, Collection<AffinityRule> affinityRules, Map<PTOperator, PTContainer> operatorContainerMap)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java:472]: Would it make sense to maintain the relationship in PhysicalPlan instead of adding to each container?

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 23 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: How do we handle this with dynamic plan changes?

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java (hunk 22 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: I have put it inside loop so that node reports get updated periodically. This helps in keeping latest available resources in node reports so that on container kill/restart, we can allocate nodes based on latest available resources.

File: engine/src/main/java/com/datatorrent/stram/BlacklistBasedResourceRequestHandler.java (hunk 21 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/BlacklistBasedResourceRequestHandler.java]: Fixed it.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 24 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: This check cannot be done in logical Plan validation. Since, 2 operators could be assigned to same container even if there is not locality constraint. This may occur if maxContainer value is set less than the needed number of containers needed for the application.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 25 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java:472]: Both are possible ways of doing it. This is more consistent with how node locality is handled i.e. PTOperator contains list of operators that should be co-allocated. Advantage of having the anti-affinity list of containers in each container is it makes it easy to assign host to container requests based on anti-affinity list.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 24 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: maxContainer is deprecated, so the check won't be needed then?

File: engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java (hunk 22 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java]: It has to be in the loop, but no need to call it multiple times. The millis are used in several places.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 24 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Oh, yes then the check won't be needed. Will remove it.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 23 above)
Comment (ishark) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Great point. Did not handle it earlier. 
Removed clearing of inlinePrefs and localityPrefs. These preference are applied again on dynamic plan changes. So affinity will be set again on dynamic changes. As for anti-affinity, added population of antiAffinityPreferences for containers on re-allocation.

Added a test case for the same.

**Comment by ishark** [comment]
Have addressed review comments and updated diff.
Tests ran fine locally.

**Comment by tweise** [comment]
@ishark please resolve the merge conflict

File: api/src/main/java/com/datatorrent/api/StringCodec.java [hunk 26]
```diff
@@ -377,4 +384,43 @@ public String toString(Class<? extends T> clazz)
...
+    @Override
+    public T fromString(String string)
+    {
+      try {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.configure(DeserializationConfig.Feature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        ObjectReader reader = mapper.reader(clazz);
+        return reader.readValue(string);
+      } catch (IOException e) {
+        Throwables.propagate(e);
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StringCodec.java]: Exception propagation. No need for subsequent return null

Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StringCodec.java]: Would be nice to fix the other occurrences in this file also and remove usage of DTThrowable.

**Comment by tweise** [comment]
Are the affinity sets updated when containers are added or removed due to dynamic partitioning? If they are not, please create a follow up JIRA so we can address it later.

**Comment by ishark** [comment]
I have added reassigning of anti-affinity sets for containers in assignContainers after repartitioning is done. But will create a follow up Jira to validate dynamic update scenarios.
